jax/_src/lax/control_flow/loops.py

```diff
@@ -1666,9 +1666,14 @@ def fori_loop(lower, upper, body_fun, init_val):
   lower_dtype = dtypes.canonicalize_dtype(lax.dtype(lower))
   upper_dtype = dtypes.canonicalize_dtype(lax.dtype(upper))
   if lower_dtype != upper_dtype:
-    msg = ("lower and upper arguments to fori_loop must have equal types, "
-           "got {} and {}")
-    raise TypeError(msg.format(lower_dtype.name, upper_dtype.name))
+    # Check for scalar integer inputs in case of type mismatch
+    if dtypes.is_weakly_typed(lower) and np.issubdtype(lower_dtype, np.integer):
```

Comment:
Wouldn't this be problematic in the case that both are weakly typed and upper is the one with lower precision?

Reply:
Yes, but that's part of the compromise we've made in allowing python scalar integers to implicitly promote to any strong integer type.

Reply:
Oh, nvm, I see what you're saying...

Reply:
Man this is tough to do cleanly in a way that preserves staticness of inputs (required for choosing `scan` vs. `while` below). i.e. we can't use `lax.convert_element_type` because of omnistaging